### ISSUE TITLE
Workers fail to accept job commands after brief redis connection loss

### ISSUE THREAD

**Issue by buroa** [rq/rq#1836]
I'm not sure why, but I think it has to do with the pubsub channel. The only way to "fix" the issue presently is to basically restart the worker.

**Comment by lowercase00** [comment]
How brief is the connection loss?

**Comment by buroa** [comment]
Roughly less than a minute & this is how I setup a redis connection on the worker:

```RETRY = Retry(ExponentialBackoff(10, 1), 12)
RETRY_ERRORS = [TimeoutError, ConnectionError, ConnectionRefusedError]

def get_redis_conn():
    redis_url = os.getenv("REDIS_URL", "redis://localhost:6379")
    return redis.from_url(
        redis_url,
        retry=RETRY,
        retry_on_error=RETRY_ERRORS,
        socket_timeout=(60 * 2) + 5,
        health_check_interval=15,
    )
```

**Comment by lowercase00** [comment]
Ok, I was asking because for longer connection losses it seems there's a bug where the worker key is expired but the Worker keeps the heartbeats, so I was wondering if that was because of that. But considering the connection loss is brief, this is most likely related to the `pubsub` subscription.

Did you have the chance to check whether the worker still has the subscription to the channel (after it reestablishes the connection)?

**Comment by buroa** [comment]
Looks like none are listening. I just replicated the issue in a local development environment and same thing.

You can replicate by:

1. create a rq worker
2. shut down redis for 15 seconds and then start it
3. create a long running job + attempt to stop it: it won't stop

```127.0.0.1:6379> pubsub channels
(empty array)
```

**Comment by lowercase00** [comment]
Thanks for that @buroa, wonder if the `redis-py` exception `ConnectionLost` is silent because `pubsub` runs on a separate thread. Even in this case, quickly going through `redis-py` it seems there's a hook that should be reconnecting the channels in case of lost connections:

```python
...
    def on_connect(self, connection):
        "Re-subscribe to any channels and patterns previously subscribed to"
```

This is registered here:

```python
...
# register a callback that re-subscribes to any channels we
# were listening to when we were disconnected
self.connection.register_connect_callback(self.on_connect)
```

Anyways, will take a look at this, thanks for the report.

**Comment by buroa** [comment]
Thanks @lowercase00, let me know if you need help. I'm debugging on my end too.

**Comment by NefixEstrada** [comment]
Hello,

We've encountered the same bug. We've been investigating why it happens and why there's no reconnection, and I think we've found the reason:

1. The subscription connection is issued when the worker starts to work, more specifically in the `subscribe` function: https://github.com/rq/rq/blob/master/rq/worker.py#L914
2. This calls the `pubsub` function in the Redis connection and then does the actual subscription using `pubsub.subscribe`, which is the following `redis-py` function: https://github.com/redis/redis-py/blob/e3de026a90ef2cc35a5b68934029a0ef2a5b2f53/redis/client.py#L876
3. This one calls the `execute_command` function of the `PubSub` class (https://github.com/redis/redis-py/blob/e3de026a90ef2cc35a5b68934029a0ef2a5b2f53/redis/client.py#L713)
4. This function, checks if there's a connection, and if there's not, does the reconnection using `self.on_connect` (https://github.com/redis/redis-py/blob/e3de026a90ef2cc35a5b68934029a0ef2a5b2f53/redis/client.py#L726)

This is only executed when the worker is bootstrapped, but if the connection drops, there's never a reconnection attempt.

This is my guess on what's happening


CC: @jvinolas

**Comment by bfreskura** [comment]
@buroa Have you found some workarounds besides restarting the worker?

**Comment by buroa** [comment]
> @buroa Have you found some workarounds besides restarting the worker?

Apologies. I no longer utilize this project as I left my company where I was using it.

**Comment by fcharlier** [comment]
Hi,

I stumbled on the same issue after we had our Redis server restart regularly and I noticed the workers stopped answering to shutdown commands but still continued accepting and processing jobs.

TL;DR, if the connection to the Redis server drops for any reason, an exception is raised and [caught in the worker retry mechanism](https://github.com/rq/rq/blob/master/rq/worker.py#L1048-L1054)  the connection is established again at the next `heartbeat()` call **but the pubsub handlers are not reinstalled**.

Overriding the connection class as following using redis-py's retry mechanism allows to work-around this issue when run with `rq worker --connection-class redis_with_retries.RedisWithRetries […]`

```python
import socket

from redis import Redis
from redis.backoff import ExponentialBackoff
from redis.retry import Retry
import redis.exceptions


class RedisWithRetries(Redis):
    @classmethod
    def from_url(cls, url: str, **kwargs) -> None:
        kwargs.update(
            {
                "retry": Retry(backoff=ExponentialBackoff(2.0, 0.1), retries=-1),
                "retry_on_error": [
                    ConnectionError,
                    ConnectionResetError,
                    ConnectionRefusedError,
                    redis.exceptions.ConnectionError,
                    redis.exceptions.TimeoutError,
                    socket.gaierror,
                    socket.timeout,
                ],
            }
        )

        return super().from_url(url, **kwargs)
```

However it has two drawbacks:

1.  For it to work correctly I had to set the number of retries to -1 (always retry) (if not, after a certain amount of retries, it would enter RQ's retry mechanism and we're back to missing pubsub handlers) and since there's no logging occurring in redis-py's retry handler, a configuration mistake or other permanent error would be difficult to diagnose.
2. Using this connection class override basically bypasses RQ's retry mechanism for the worker.

From here, I see two options:

- Either replace RQ's worker retry mechanism with redis-py's retry mechanism by default in RQ
- Or modify RQ's worker retry mechanism to re-install the pubsub handlers after an exception occurs.

What do you think ?
